### DISCUSSION TITLE
Recommended usage for personal class/style files

### DISCUSSION THREAD

**Discussion by librarianmage** [tectonic-typesetting/tectonic#998] [Q&A]
I have a personal assignment class file I use for all of my class assignments. Is there a recommended way to reuse this file in such a way that I don’t need to copy it to every assignment folder I have?

**Comment by pkgw** [comment]
In my workflows, I generally just do the lame thing and copy the files between documents. It's certainly not elegant, but it gets the job done.

For one-off builds (the "V1" interface or `tectonic -X compile`), you can use the [`-Z search-path`][1] unstable option, but it's a bit awkward to use. And that's intentional: a design goal of Tectonic is reproducibility in its document builds, and having the build depend on external resources breaks that.

The idea is that the "V2" document model will eventually support "workspaces" of multiple documents that can share resources, addressing this kind of use case. Unfortunately, someone needs to sit down and implement that feature, and that hasn't happened yet.

If you're not on Windows, you might be able to reduce duplication a bit by using symbolic links. Depending on your circumstances, that might not feel like much of an improvement over just copying files around.

One final option that could possibly help is putting your common files into a Git repository, and then using Git's "submodule" functionality to pull them into a parent repo for each document. This doesn't reduce duplication, but it does add a layer of precision in managing your shared resources.

[1]: https://tectonic-typesetting.github.io/book/latest/v2cli/compile.html#unstable-options

I have a few other ideas for the future that I hope will address this common need, but I want to implement a solution that keeps document builds reproducible and precisely specified. That basically requires moving towards a model like Rust or NPM where each package specifies dependencies that are managed in some packaging framework, which unfortunately is a relatively large engineering lift. The good news is that Tectonic can probably piggyback on some modern package management system (Cargo, NPM, Conda) rather than trying to implement a whole new thing itself.
